### PR TITLE
GH-2639 - classloader assertions implementation.

### PR DESCRIPTION
**Pull request by ascopes**
This implements the classes defined in #2645.

The PR currently a work-in-progress, and I will be implementing it over the next week or so when I find time. This is an initial head start.

Once the PR is ready for review, I will update this description with something more descriptive.

**This PR requires GH-2645 to be merged first. This includes that PR's content as it is dependent on the definitions within that PR.**

----

TODOs (for my reference):

- [x] Examples
- [x] Tests for Assertion classes
- [x] Tests for error messages
- [x] Tests for any internal API changes
- [x] Verify binary compatibility with JDK-8 through JDK-19

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractClassLoaderAssert.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+import java.util.Enumeration;
+import org.assertj.core.internal.Failures;
+import org.assertj.core.internal.Objects;
+import org.assertj.core.util.CheckReturnValue;
+
+/**
+ * Base class for all implementations of assertions for {@link ClassLoader class loaders}.
+ *
+ * @param <SELF> the "self" type of this assertion class. Please read &quot;<a
+ *               href="http://bit.ly/1IZIRcY" target="_blank">Emulating 'self types' using Java
```
Comment (vlsi) [inline src/main/java/org/assertj/core/api/AbstractClassLoaderAssert.java:34]: Would it be better to put the full URL here?

Comment (ascopes) [reply src/main/java/org/assertj/core/api/AbstractClassLoaderAssert.java:34]: Yeah, I agree. This was reusing the documentation headers from existing code in this project, but I can update that. Good shout

Comment (ascopes) [reply src/main/java/org/assertj/core/api/AbstractClassLoaderAssert.java]: Updated

**Comment by ascopes** [comment]
The compatibility error here shouldn't be an issue.

```
Error:  Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.15.7:cmp (default-cli) on project assertj-core: There is at least one incompatibility: org.assertj.core.util.Streams.Streams():CONSTRUCTOR_LESS_ACCESSIBLE,org.assertj.core.util.Streams:CLASS_NOW_FINAL -> [Help 1]
```

This is caused by me making the `Streams` class `final` when I added additional functionality to it. No one should be making instances of that class since the functionality is static. I also made the constructor private to enforce that it is a static-only class, which is where the second issue came from.

**Comment by ascopes** [comment]
@scordio out of curiosity, is this just marked as draft due to the binary compatibility script failure/the dependent base PR? Apart from that this should be ready for review hopefully :-)

**Comment by scordio** [comment]
@ascopes I think I've done it only because you mentioned it being WIP in the description. No worries, let's switch it to ready and I'll be on it during the week!

**Comment by ascopes** [comment]
Unless there is a timeframe for getting this done before then, I will take a look at fixing this at the weekend. Should be able to just drop the commits already on master as they just duplicate that.

Thanks!

**Comment by scordio** [comment]
No worries, take your time!

**Comment by ascopes** [comment]
Should be fixed now!

**Comment by scordio** [comment]
@ascopes we plan to release a new version in the upcoming weeks and I'm not sure I can manage manage to finalize this PR in time.

As the base branch evolved and there are now several conflicts, I am going to revert ce3ee9b57c8339d9b3f9f6b6e96421ce1b5bc133 and I will create a new branch with those changes and the ones in this PR, keeping you as the author of course.

Sorry for the delay and thanks a lot for your patience 🙂

**Comment by ascopes** [comment]
> @ascopes we plan to release a new version in the upcoming weeks and I'm not sure I can manage manage to finalize this PR in time.
> 
> As the base branch evolved and there are now several conflicts, I am going to revert ce3ee9b57c8339d9b3f9f6b6e96421ce1b5bc133 and I will create a new branch with those changes and the ones in this PR, keeping you as the author of course.
> 
> Sorry for the delay and thanks a lot for your patience 🙂 

No worries at all! Thanks for the update :-)

**Comment by scordio** [comment]
> I am going to revert [ce3ee9b](https://github.com/assertj/assertj/commit/ce3ee9b57c8339d9b3f9f6b6e96421ce1b5bc133) and I will create a new branch with those changes and the ones in this PR, keeping you as the author of course.

Closing in favor of #2973.

**Pull request closed without merging**